File: src/main/java/it/unitn/dbtrento/spark/writer/SparkWriter.java
package it.unitn.dbtrento.spark.writer;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.HashMap;
import java.util.Map;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.FileUtil;
import org.apache.hadoop.fs.Path;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.SaveMode;
import org.apache.spark.sql.SparkSession;

import it.unitn.dbtrento.spark.utils.OutputFormat;

public class SparkWriter {

  public static <T> boolean write(SparkSession spark, Dataset<T> data, boolean hasHeader,
      String outputPath, OutputFormat outputFormat) {
    Map<String, String> options = new HashMap<>();
    options.put("header", String.valueOf(hasHeader));
    return write(spark, data, options, outputPath, outputFormat);
  }

  public static <T> boolean write(SparkSession spark, Dataset<T> data, boolean hasHeader,
      String outputPath, OutputFormat outputFormat, String outputFile) {
    Map<String, String> options = new HashMap<>();
    options.put("header", String.valueOf(hasHeader));
    return write(spark, data, options, outputPath, outputFormat, outputFile);
  }

  public static <T> boolean write(SparkSession spark, Dataset<T> data, Map<String, String> options,
      String outputPath, OutputFormat outputFormat) {
    return write(spark, data, options, outputPath, outputFormat, "output");
  }

  public static <T> boolean write(SparkSession spark, Dataset<T> data, Map<String, String> options,
      String outputPath, OutputFormat outputFormat, String outputFile) {
    if (data == null) {
      System.out.println("Not able to write the data...");
      return false;
    } else if (outputPath == null) {
      System.out.println("Not able to write the data...");
      return false;
    } else if (outputPath.isEmpty()) {
      System.out.println("Not able to write the data...");
      return false;
    }
    try {
      switch (outputFormat) {
        case CSV:
          data.write().options(options).mode(SaveMode.Overwrite).csv(outputPath);
          return true;
        case PARQUET:
          data.write().options(options).mode(SaveMode.Overwrite).save(outputPath);
          return true;
        case SINGLE_CSV:
          data.coalesce(1).write().options(options).mode(SaveMode.Overwrite)
              .csv(outputPath + "/partial/");
          try {
            FileUtil.copyMerge(
                FileSystem.get(new URI(outputPath + "/partial/"), new Configuration()),
                new Path(outputPath + "/partial/"),
                FileSystem.get(new URI(outputPath + "/" + outputFile + ".csv"),
                    new Configuration()),
                new Path(outputPath + "/" + outputFile + ".csv"), true,
                spark.sparkContext().hadoopConfiguration(), null);
          } catch (IllegalArgumentException | IOException | URISyntaxException e) {
            System.err.println("Error " + e.getMessage() + "while writing the data");
            e.printStackTrace();
          }
          return true;
        default:
          return false;
      }
    } catch (UnsupportedOperationException e) {
      System.err
          .printf("%s while writing the data to %s in %s format.\n", e.getMessage(), outputPath,
              outputFormat.toString());
    }
    return false;
  }
}

File: src/main/scala/eu/unitn/disi/db/spark/io/SparkWriter.scala
package eu.unitn.disi.db.spark.io

import java.net.URI

import eu.unitn.disi.db.spark.utils.Utils
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.{FileSystem, FileUtil, Path}
import org.slf4j.{Logger, LoggerFactory}

import org.apache.spark.sql.{DataFrameWriter, Dataset, SaveMode, SparkSession}

object SparkWriter {

  val log: Logger = LoggerFactory.getLogger(this.getClass.getName)

  def write(spark: SparkSession,
            dataset: Dataset[_],
            header: Boolean,
            path: String,
            outputFormat: Format): Boolean = {
    write(spark, dataset, Map("header" -> header.toString), path, outputFormat)
  }

  def write(spark: SparkSession,
            dataset: Dataset[_],
            options: java.util.Map[String, String],
            path: String,
            outputFormat: Format): Boolean = {
    write(spark,
      dataset,
      Utils.javaToScalaMap[String, String](options),
      path,
      outputFormat)
  }

  def write(spark: SparkSession,
            dataset: Dataset[_],
            options: java.util.Map[String, String],
            path: String,
            outputFormat: Format,
            filename: String): Boolean = {
    write(spark,
      dataset,
      Utils.javaToScalaMap[String, String](options),
      path,
      outputFormat,
      filename)
  }

  def write(spark: SparkSession,
            dataset: Dataset[_],
            options: Map[String, String],
            path: String,
            format: Format,
            filename: String = "output"): Boolean = {
    if (dataset == null || path == null || path.isEmpty) {
      log.debug("Not able to write the data...")
      return false
    }

    val writer: DataFrameWriter[_] = (format match {
      case Format.SINGLE_CSV => dataset.coalesce(1).write
      case _ => dataset.write
    }).options(Utils.addSeparatorToOptions(options, format))
      .mode(SaveMode.Overwrite)

    var returnValue: Boolean = false

    try {
      returnValue = format match {
        case Format.CSV | Format.SSV | Format.TSV =>
          writer.csv(path)
          true
        case Format.PARQUET =>
          writer.parquet(path)
          true
        case Format.SINGLE_CSV =>
          writer.csv(path + "/partial/")
          try {
            val fullPath: String = if (filename.isEmpty || filename == null) {
              path
            } else {
              path + "/" + filename + ".csv"
            }

            FileUtil.copyMerge(
              FileSystem.get(new URI(path + "/partial/"), new Configuration),
              new Path(path + "/partial/"),
              FileSystem.get(new URI(fullPath), new Configuration),
              new Path(fullPath),
              true,
              spark.sparkContext.hadoopConfiguration,
              null
            )
            true
          } catch {
            case e: Throwable =>
              log.debug(s"Error ${e.getMessage} while writing the data!")
              false
          }
        case Format.JSON =>
          writer.json(path)
          true
        case _ => false
      }
    } catch {
      case e: UnsupportedOperationException =>
        log.error(
          s"Error ${e.getMessage} while writing the data to $path/$filename in ${format.toString} format.")
    }
    returnValue
  }
}
